File: frontend/lib/api/courses.ts
import { apiGet } from './client';

export type ApiCredit = {
  lecture: number;
  lab: number;
  credit: number;
  au: number;
  raw: string;
};

export type ApiOfferedSemester = 'S' | 'F';

export type ApiCourse = {
  id: string;
  courseCode: string;
  courseName: string;
  courseNameEn: string | null;
  category: string;
  sectors: string[];
  offeredSemesters: ApiOfferedSemester[];
  credit: ApiCredit;
  prerequisites: string[];
  isKeyCourse: boolean;
  level: number | null;
  matched: boolean;
};

export type ApiCourseCategory = {
  category: string;
  nameEn: string;
  order: number;
  courseCount: number;
};

export type ApiCourseSector = {
  sector: string;
  nameEn: string;
  order: number;
  courseCount: number;
};

export type ListCoursesParams = {
  q?: string;
  category?: string;
  sector?: string;
  offeredSemester?: ApiOfferedSemester;
  isKeyCourse?: boolean;
  level?: number;
  includePrerequisites?: boolean;
};

export function listCourses(params?: ListCoursesParams) {
  return apiGet<ApiCourse[]>('/courses', params);
}

export function getCourse(courseCode: string) {
  return apiGet<ApiCourse>(`/courses/${courseCode}`);
}

export function listCourseCategories() {
  return apiGet<ApiCourseCategory[]>('/courses/categories');
}

export function listCourseSectors() {
  return apiGet<ApiCourseSector[]>('/courses/sectors');
}

File: backend/src/fastapi_app/schemas/courses.py
from pydantic import BaseModel, ConfigDict, Field

from db.models.course import CourseCredit


class CourseDTO(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    course_code: str = Field(serialization_alias="courseCode")
    course_name: str = Field(serialization_alias="courseName")
    course_name_en: str | None = Field(serialization_alias="courseNameEn")
    description: str | None = None
    category: str
    sectors: list[str]
    offered_semesters: list[str] = Field(serialization_alias="offeredSemesters")
    credit: CourseCredit
    prerequisites: list[str]
    is_key_course: bool = Field(serialization_alias="isKeyCourse")
    level: int | None
    matched: bool = True


class CourseCategoryDTO(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    category: str
    name_en: str = Field(serialization_alias="nameEn")
    order: int
    course_count: int = Field(serialization_alias="courseCount")


class CourseSectorDTO(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    sector: str
    name_en: str = Field(serialization_alias="nameEn")
    order: int
    course_count: int = Field(serialization_alias="courseCount")
